ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,807 @@
+          // detection.
+          chunkException = new RouterException("Response deserialization received an unexpected error", e,
+              RouterErrorCode.UnexpectedInternalError);
+          chunkOperationTracker.onResponse(getRequestInfo.replicaId, false);
+        }
+      }
+      checkAndMaybeComplete();
+    }
+
+    /**
+     * Process the GetResponse extracted from a {@link ResponseInfo}
+     * @param getRequestInfo the associated {@link RequestInfo} for which this response was received.
+     * @param getResponse the {@link GetResponse} extracted from the {@link ResponseInfo}
+     * @throws IOException if there is an error during deserialization of the GetResponse.
+     * @throws MessageFormatException if there is an error during deserialization of the GetResponse.
+     */
+    private void processGetBlobResponse(GetRequestInfo getRequestInfo, GetResponse getResponse)
+        throws IOException, MessageFormatException {
+      ServerErrorCode getError = getResponse.getError();
+      if (getError == ServerErrorCode.No_Error) {
+        int partitionsInResponse = getResponse.getPartitionResponseInfoList().size();
+        // Each get request issued by the router is for a single blob.
+        if (partitionsInResponse != 1) {
+          chunkException = new RouterException("Unexpected number of partition responses, expected: 1, " +
+              "received: " + partitionsInResponse, RouterErrorCode.UnexpectedInternalError);
+        } else {
+          getError = getResponse.getPartitionResponseInfoList().get(0).getErrorCode();
+          responseHandler.onRequestResponseError(getRequestInfo.replicaId, getError);
+          if (getError == ServerErrorCode.No_Error) {
+            handleBody(getResponse.getInputStream());
+            chunkOperationTracker.onResponse(getRequestInfo.replicaId, true);
+          } else {
+            // process and set the most relevant exception.
+            processServerError(getError);
+            if (getError == ServerErrorCode.Blob_Deleted || getError == ServerErrorCode.Blob_Expired) {
+              // this is a successful response and one that completes the operation regardless of whether the
+              // success target has been reached or not.
+              chunkCompleted = true;
+            } else {
+              chunkOperationTracker.onResponse(getRequestInfo.replicaId, false);
+            }
+          }
+        }
+      } else {
+        responseHandler.onRequestResponseError(getRequestInfo.replicaId, getError);
+        chunkOperationTracker.onResponse(getRequestInfo.replicaId, false);
+      }
+    }
+
+    /**
+     * Process the given {@link ServerErrorCode} and set operation status accordingly.
+     * @param errorCode the {@link ServerErrorCode} to process.
+     */
+    void processServerError(ServerErrorCode errorCode) {
+      setChunkException(new RouterException("Server returned: " + errorCode, RouterErrorCode.UnexpectedInternalError));
```

Comment:
Yes. Also see the documentation for this method in `FirstChunk`, which I have now also added to the docs here.

If a data chunk is being retrieved, that means the first chunk was retrieved successfully, and the blob is "good", so any error is unexpected. We can't return a blob deleted error for a chunk at this point, for example, as the operation callback was already returned with "blob found".